pkg/secrets/externaldriver/externaldriver.go

```diff
@@ -0,0 +1,159 @@
+package externaldriver
+
+import (
+	"bytes"
+	"context"
+	"os"
+	"os/exec"
+	"sort"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+)
+
+var (
+
+	// errMissingConfig indicates that one or more of the external actions are not configured
+	errMissingConfig = errors.New("missing config value")
+
+	// errNoSecretData indicates that there is not data associated with an id
+	errNoSecretData = errors.New("no secret data with ID")
+
+	// errInvalidKey indicates that something about your key is wrong
+	errInvalidKey = errors.New("invalid key")
+)
+
+type driverConfig struct {
+	// DeleteCommand contains a shell command that deletes a secret.
+	// The secret id is provided as environment variable SECRET_ID
+	DeleteCommand string `mapstructure:"delete"`
+	// ListCommand contains a shell command that lists all secrets.
+	// The output is expected to be one id per line
+	ListCommand string `mapstructure:"list"`
+	// LookupCommand contains a shell command that retrieves a secret.
+	// The secret id is provided as environment variable SECRET_ID
+	LookupCommand string `mapstructure:"lookup"`
+	// StoreCommand contains a shell command that stores a secret.
+	// The secret id is provided as environment variable SECRET_ID
+	// The secret value itself is provied over stdin
+	StoreCommand string `mapstructure:"store"`
+}
+
+func (cfg *driverConfig) ParseOpts(opts map[string]string) error {
+	if err := mapstructure.Decode(opts, cfg); err != nil {
+		return err
+	}
+	if cfg.DeleteCommand == "" ||
+		cfg.ListCommand == "" ||
+		cfg.LookupCommand == "" ||
+		cfg.StoreCommand == "" {
+
+		return errMissingConfig
+	}
+	return nil
+}
+
+// Driver is the passdriver object
+type Driver struct {
+	driverConfig
+}
+
+// NewDriver creates a new secret driver.
+func NewDriver(opts map[string]string) (*Driver, error) {
+	cfg := &driverConfig{}
+	if err := cfg.ParseOpts(opts); err != nil {
+		return nil, err
+	}
+
+	driver := &Driver{
+		driverConfig: *cfg,
+	}
+
+	return driver, nil
+}
+
+// List returns all secret IDs
+func (d *Driver) List() (secrets []string, err error) {
+	cmd := exec.CommandContext(context.TODO(), "/bin/sh", "-c", d.ListCommand)
```

Comment:
`cmd := exec.CommandContext(context.TODO(),` _**"/bin/sh", "-c",**_ `d.ListCommand)`

Is this going to bite us on windows?

Reply:
This driver should be built to be non windows.  I would guess.

Reply:
Why do we need `/bin/sh`?  Couldn't we use `exec.Command(d.ListCommand[0], d.ListCommand[1:]...)`?

Reply:
@vrothberg I used a shell on purpuse to allow simple shell expressions and the use of environment variables. Look at how the tests are setup. For example "store" is implemented using this expression: `cat - > /tmp/test/${SECRET_ID}`. I think in real usecases it will also be usefull to for example pipe some json result to `jq` to extract the actual secret value etc.